### DISCUSSION TITLE
configuration issue with sensor groups

### DISCUSSION THREAD

**Discussion by ranug14** [slipx06/sunsynk-power-flow-card#402] [Huawei / iStore]
Hello,

first of all, thanks a lot for you effort and patience. I am really grateful and always amazed at what is possible thanks to the community.

I'm following the instructions on https://slipx06.github.io/sunsynk-power-flow-card/examples/huawei.html to get the card running with my Huawei setup (SUN2000-10KTL-M1 inverter with Luna2000 Battery). Unfortunately I stuck at the point where I have to create the sensor groups. Starting with the creation of group "Sunsynk Card - AUX - Energy Daily" I have no clue what members to add to this group. In the purpose column its stated "Group should be populated with the daily energy sensors of all the devices... ", but what are these devices?

Thanks a lot,
ranug

**Comment by slipx06** [comment]
Hi @Roving-Ronin was working on this and can help you

**Comment by Roving-Ronin** [comment]
@ranug14 

"Sunsynk Card - AUX - Energy Daily" is the energy_meter (enhanced version of the HA utility_meter) that is tracking the kWh used each day by the devices in the SunSynk cards 'AUX' version (the blue-ish section in the top right of example card below). 

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/108674933/ed84cc0c-5b1b-449b-9dbb-1d6fe865163c)

Highlighed by the red box is the value of this energy meter, whilst in yellow below it is the current 'Active Power' of the same devices. In regards 'what devices' to add, this is entirely up to yourself. By default the 'AUX' section of the card is for things like monitoring an auxiliary power feed (i.e. generator set) but given most people won't have this, for the Huawei / iStore examples it has been repurposed to allow people to so their major essential items in the house. In the example above, the users main critical items that use the most power are his 'ITC Infrastructure' (i.e. NAS, Router, LAN Switches etc) and the other is his 'IT - Desktops'.   These where existing groups that he had to provide a 'sum' value of those device types Active Power utilisation, that are showing the 421W and 159W values.

To get the 'Sunsynk Card - AUX - Active Power' sensor that shows the value for these two other groups, he's then created group of this name...   Steps:   Settings / Devices & Services / Helpers Tab.  Then 'Create Helper' and choose 'Group', then 'Sensor Group' and enter the name 'Sunsynk Card - AUX - Active Power', then add the sensors for whatever devices POWER entities you want to track, and finish by setting Type to 'SUM'.

You now have a group giving the POWER of the devices, but you now need to convert POWER into ENERGY (i.e. Watts to kWh). To do this you now need to create another helper.  Steps:   Settings / Devices & Services / Helpers Tab.  Then 'Create Helper' and choose 'Integration - Riemann sum integral', and complete it as per:

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/108674933/bf7ceb5b-128f-43a1-88c5-67c2e742cff2)

OK you now have a sensor providing the ENERGY used by all the devices, whose POWER sensors you added to the AUX Group. This sensor however will keep on adding the power used as an ever increasing total (i.e. a 'lifetime' value), and you need a sensor that provides you with the daily amount of Energy (kWh) that AUX group used for the SunSync card. 

So now you need to create a final helper, that is a utility meter (or if your into YAML you can create it using the energy_meter integration, that will allow you to also specify the current energy cost and have it automatically create another sensors that tracks the daily cost $ for the energy used by the AUX group). To create a simple utility meter though.  Steps:   Settings / Devices & Services / Helpers Tab.  Then 'Create Helper' and choose 'Utility Meter'  and configure as per below:

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/108674933/724c9c72-7153-4a97-b7aa-3fd8f7b8500e)

In the example yaml code, for entering into the card in the dashboard, the utility meter you've just created is defined with the following that will have it appear in the red highlighted box: 
```
entities:
  day_aux_energy: sensor.sunsynk_card_aux_energy_daily
```

Whilst the POWER reading, that appears in the Yellow highlighted box is already defined at:
```
entities:
  aux_power_166: sensor.sunsynk_card_aux_active_power
```


Before you ask, the 'Sunsynk Card - Non Essential - Active Power' is exactly the same, but only having the first step of creating a helper sensor group to sum up the non-essential items you choose to show at the bottom right of the card. The Hauwei / iStore examples being the AirCon and EV Charger. So if that reflects the devices you'd choose for there, then you'd create this 'Sunsynk Card - Non Essential - Active Power' (sensor group) and add the Power sensors of your AirCon/HVAC and EV Charger to it, and also set it to show the SUM of those sensors. 
Given the SunSynk card doesn't have space to show an energy sensor, with the daily kWh used by these Non-Essential Devices, there's no need to create a Riemann Sum or Utility Meter sensor. 

Not needed or relaated to the SunSynk card, you'll probably have a utility (or better energy_meter) setup for these power hungry devices to track their Daily/Weekly/Monthly/Yearly Energy (kWh) consumption though :-)

**Comment by ranug14** [comment]
Thank you very much, very detailed instructions. I've understood everything, but I'm still not sure whether I prefer to classify the future EV charger as essential. But I'll just try it out when it's installed.

**Reply by Roving-Ronin** [reply]
@ranug14 as mentioned above, the AirCon / HVAC and EV Charger are in the 'Non Essential' section, at the bottom of the card.
As defined by:

```
  nonessential_power: sensor.sunsynk_card_non_essential_active_power
  non_essential_load1: sensor.hvac_active_power
  non_essential_load2: sensor.ev_charger_active_power
```

The first line being the sensor (sum of) group, that I mentioned above creating, whilst the bottom two lines are the 'Active Power' sensors for the AirCon and EV Charger (just update them to whatever your actual sensors are).

**Comment by ranug14** [comment]
Hi, I can't find the reason why my sensor.inverters_active_power is unavailable. The sensor in the huawei_derived_sensors.yaml file is normally set by {% set inverter_1_active_power = states('sensor.inverter_active_power') | float(default=0) %}. 

sensor.inverter_active_power have a valid value but inverter_1_active_power  is unavailable. The Inverter String sensors above in the yaml are working fine, but all below "Start: Inverters - Combined values for 2 inverters" not, execpt inverters_internal_temperature.

any idea?

**Reply by Roving-Ronin** [reply]
@ranug14 

inverter_1_active_power is NOT an entity in HA that you can lookup / get the status of. If you look at the yaml for this sensor below, you can see that:

        {% set inverter_1_active_power = states('sensor.inverter_active_power') | float(default=0) %}
 and
        {% set inverter_2_active_power = states('sensor.inverter_active_power_2') | float(default=0) %}

are defining a variable name / alias for use later in the state formula, with inverter_1_active_power and  inverter_2_active_power being those alias and their value is then set by calling the state of sensor.inverter_active_power / sensor.inverter_active_power_2.

The rest of the state formula is then set around seeing if there's 1 or 2 inverters and from that working out how to sum them up.

```
  # Provides the combined current Active Power of both inverters, in Watts.
  - sensor:
    - name: "Inverters - Active Power"
      unique_id: inverters_active_power
      unit_of_measurement: "W"
      device_class: "power"
      state_class: measurement
      state: >-
        {% set inverter_1_active_power = states('sensor.inverter_active_power') | float(default=0) %}
        {% set inverter_2_active_power = states('sensor.inverter_active_power_2') | float(default=0) %}

        {% if states('sensor.inverter_active_power_2') == 'unknown' %}
          {{ inverter_1_active_power }}
        {% else %}
          {% set count = 0 %}
          {% set sum = 0 %}

          {% if states('sensor.inverter_active_power') != 'unknown' %}
            {% set sum = sum + inverter_1_active_power %}
            {% set count = count + 1 %}
          {% endif %}

          {% if states('sensor.inverter_active_power_2') != 'unknown' %}
            {% set sum = sum + inverter_2_active_power %}
            {% set count = count + 1 %}
          {% endif %}

          {% if count > 0 %}
            {{ sum | round(1) }}
          {% else %}
            N/A
          {% endif %}
        {% endif %}
      availability: >-
        {{  states('sensor.inverter_active_power')   | is_number
        and states('sensor.inverter_active_power_2') | is_number }}
```

If you paste the following into Developers Tools ---> Template, you should see the values for the individual inverters. If you only have 1 inverter you should see a numerical value for it, and 'unknown' for the second value. If you have two inverters and don't see two numerical values, I'd start checking the names assigned to the inverters (and batteries etc) by the WLCRS Huawei Integration. 

```
{{ states('sensor.inverter_active_power') }}
{{ states('sensor.inverter_active_power_2') }}
```

If you installed WLCRS integration a while ago, they changed the default entity naming format in late 2023 / ~v1.3.0. Prior to that (iirc) it was something like sensor.inverter_1_active_power and sensor.inverter_2_active_power was the entity name it created, but that was changed to the current format of sensor.<inverter/battery etc>_function  and then no number for the first and then _2, _3 etc for additional units.

**Comment by ranug14** [comment]
Thanks for pointing me to the template function, really helpful. I pasted the complete yaml code of inverters_active_power and I can see that it is working as expected, there is a valid value at the end. 

But the availability (last 2 lines) is set to FALSE because I only have one inverter and the AND condition leads to false. Shouldn't it actually be OR to get TRUE if at least one of the inverters delivers correct values?

Currently {{ states('sensor.inverters_active_power') }} returns the value unavailable.

**Reply by Roving-Ronin** [reply]
@ranug14 

Sorry my bad, I might have to rewrite some of the other sensors, as I thought they would work if you had 1 or 2 inverters. 
I only have my own site (2 x M1 6kW inverters and NO battery) and another site 5000km away (2 x L1 5kW and a 10kW LUNA ESS) another gent generously allows me access to, to test... alas no 1 or 3 inverter, or 2 or more battery sites to test against.

Would you mind trying this updated sensor below and see how it responds ?

```
- platform: template
  sensors:
    inverters_active_power:
      friendly_name: "Inverters - Active Power"
      unit_of_measurement: "W"
      device_class: "power"
      state_class: measurement
      value_template: >-
        {% set inverter_1_active_power = states('sensor.inverter_active_power') | float(default=0) %}
        {% set inverter_2_active_power = states('sensor.inverter_active_power_2') | float(default=0) %}
        {% set count = 0 %}
        {% set sum = 0 %}

        {% if states('sensor.inverter_active_power') != ['unknown', 'unavailable', 'none'] %}
          {% set sum = sum + inverter_1_active_power %}
          {% set count = count + 1 %}
        {% endif %}

        {% if states('sensor.inverter_active_power_2') != ['unknown', 'unavailable', 'none'] %}
          {% set sum = sum + inverter_2_active_power %}
          {% set count = count + 1 %}
        {% endif %}

        {% if count > 0 %}
          {{ sum | round(1) }}
        {% else %}
          N/A
        {% endif %}
    availability_template: >
      {% set inverter_1_available = not states('sensor.inverter_active_power') in ['unknown', 'unavailable', 'none'] %}
      {% set inverter_2_available = false %}
      {% if 'sensor.inverter_active_power_2' in states %}
        {% set inverter_2_available = not states('sensor.inverter_active_power_2') in ['unknown', 'unavailable', 'none'] %}
      {% endif %}
      {{ inverter_1_available or inverter_2_available }}

```

**Reply by ranug14** [reply]
Thats working, value and availability (true) is correct for my setup. Should I replace this part in my packages/huawei_derived_sensors.yaml? For my setup it will also work if I replace the "and" with "or" in the availability sections. Same for the other inverter sensors in the file.

**Reply by Roving-Ronin** [reply]
@ranug14 yes I'll have to go through and find what other sensors need to be updated. I'm thinking that I might just rewrite / clean the examples up and release based upon the forthcoming release of the WLCRS Huawei Integration v1.4.0 that is currently in Alpha3, given that is a major release with underlying engine rewrite and a bunch of new sensors for polling the modbus 'batteries' (plural) registers able to be enabled (disabled by default), but that also provide a 'factory' option for users with 0/1/2+ batteries to all use the same sensors in Sunsynk card without needing to add derived sensors. 
Alas derived sensors for some Inverter combined totals will still be required (lodged a feature request though to see if InverterS - Total Active Power and Tota Input Power, can be added to WLCRS.

Now that SunSynk v4.36.0 has added support for up to 4 Essential Loads (as below) when AUX is off, I'm thinking I'll also change the Huawei examples to revert to using this (currently it repurposes the AUX for this and the total Home Consumption is displayed where your screenshot shows 2.97kW (instead of the 6.65kW location your looking at).

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/108674933/f78475a7-fba0-46f7-b8fe-79ae5c888cd6)

If you can hang out, I'll update as per above... (but its already 1am where I am) :-)

**Reply by ranug14** [reply]
Easy cheesy, no hurry. work/life balance is most important :-)

**Comment by ranug14** [comment]
Now as I can put in the correct sensor values I struggle with the calculation of the house consumption. Seems like the export to grid (power_meter_active_power = 2,97) is added to the inverters_active_power (3,68) to get the house consumption. But this should be substracted.
![grafik](https://github.com/slipx06/sunsynk-power-flow-card/assets/124453844/3501da70-4265-44b0-9246-d0ac1164e49c)
